components/arm/universalrobots/ur.go

```diff
@@ -440,3 +466,84 @@ func reader(ctx context.Context, conn io.Reader, ua *URArm, onHaveData func()) e
 		}
 	}
 }
+
+const errBuiltinKinematics = "cannot use UR builtin kinematics with obstacles or interaction spaces"
+
+func (ua *URArm) useBuiltinKinematics(worldState *commonpb.WorldState, extra map[string]interface{}) (bool, error) {
+	// function to error out if trying to use world state with builtin kinematics
+	checkWorldState := func(usingBuiltinKinematics bool) (bool, error) {
+		if usingBuiltinKinematics && worldState != nil && (len(worldState.Obstacles) != 0 || len(worldState.InteractionSpaces) != 0) {
+			return false, errors.New(errBuiltinKinematics)
+		}
+		return usingBuiltinKinematics, nil
+	}
+
+	// if runtime preference is specified, obey that
+	if extra != nil {
+		if usingAtRuntime, ok := extra["builtin_kinematics"].(bool); ok {
+			return checkWorldState(usingAtRuntime)
+		}
+	}
+
+	// otherwise default to option provided at config time
+	return checkWorldState(ua.builtinKinematics)
+}
+
+// MoveToPositionURDriver uses the builtin kinematics.
+func (ua *URArm) moveWithBuiltinKinematics(ctx context.Context, pose *commonpb.Pose) error {
+	// UR5 arm takes R3 angle axis as input
+	aa := spatialmath.NewPoseFromProtobuf(pose).Orientation().AxisAngles().ToR3()
+
+	// write command to arm, need to request position in meters
+	cmd := fmt.Sprintf("movej(get_inverse_kin(p[%f,%f,%f,%f,%f,%f]), a=1.4, v=4, r=0)\r\n",
+		0.001*pose.X,
+		0.001*pose.Y,
+		0.001*pose.Z,
+		aa.X,
+		aa.Y,
+		aa.Z,
+	)
+	_, err := ua.conn.Write([]byte(cmd))
+	if err != nil {
+		return err
+	}
+
+	retried := false
+	slept := 0
+	for {
+		cur, err := ua.EndPosition(ctx, nil)
+		if err != nil {
+			return err
+		}
+		if arm.PositionGridDiff(pose, cur) <= 1.5 &&
+			arm.PositionRotationDiff(pose, cur) <= 1.0 {
+			return nil
+		}
+
+		err = ua.getAndResetRuntimeError()
+		if err != nil {
+			return err
+		}
+
+		slept += 10
+
+		if slept > 5000 && !retried {
```

Comment:
I know this is Eliot's original code, but there's got to be a better way to do this? Some positions, especially if a low speed is set, may well take more than 5 or 10 seconds to reach.

Reply:
I'm definitely open to changing this if you want, but also this code is duplicated in the `MoveToJointPositionsRadians`, s its been in production for the past few years and hasn't been a problem so far.  This might be a good task for a different ticket/PR

Reply:
Yeah this shouldn't gate the PR, just want to have this start bouncing around people's heads.

This *has* been a problem before though; there have been people using UR5 arms and getting RDK errors because it's moving slowly and this times out. The magnitude of the problem is more a function of how often people have been using the UR5 with RDK, rather than a reflection of if this is problematic.

Reply:
I added this context as a  comment on the now infamous RSDK-307 ticket.